docs/sc-memory/api/cpp/extended/agents/agents.md

```diff
@@ -0,0 +1,731 @@
+2. **Dynamic agent specification** is a specification provided in the knowledge base or initially in the code but automatically saved into the knowledge base. After that, the first agent changes the specification of the second agent, and the second agent uses the changed specification. To implement this type of specification, use the API of `ScModule` class and the API of `ScAgentBuilder` class (see [**C++ Modules**](modules.md)).
+3. **Semi-dynamic agent specification** can be obtained when the specification is provided in the knowledge base or initially in the code and appended externally (via overriding public getters).
+
+Static agent specification can be useful if you are implementing an agent in C++ for the first time or if you want to minimize the number of searches in the knowledge base. Dynamic agent specification provides the opportunity to analyze and change this specification by other agents. Semi-dynamic specification can be useful if you want to change some parts of this specification but still want the agent's calls to be quick.
+
+---
+
+## **Static agent specification**
+
+Here, the API for implementing an agent with a static specification will be discussed. This is the easiest implementation variant to understand for you. To learn about dynamic agent specification, see [**C++ Modules API**](modules.md).
+
+There are two main classes that you can use to implement an agent: `ScAgent` and `ScActionInitiatedAgent`.
+
+### **ScAgent**
+
+There is a base class for agents in C++. This class provides implemented methods to retrieve elements of the agent's specification from the knowledge base. All these methods can be overridden in your agent class.
+
+You should distinguish between an abstract sc-agent as some class of functional equivalent sc-agents described in the knowledge base and `ScAgent` as a C++ class that implements an API to work with abstract sc-agents in the knowledge base.
+
+This class can be used for all types of platform-dependent agents. The example using this class is represented below.
+
+```cpp
+// File my_agent.hpp
+#pragma once
+
+#include <sc-memory/sc_agent.hpp>
+
+// Inherit your agent class from `ScAgent` class and specify template argument 
+// as sc-event class. Here `ScEventAfterGenerateIncomingArc<
+// ScType::EdgeAccessConstPosPerm>` is type of event to which the given 
+// agent reacts.
+class MyAgent : public ScAgent<
+  ScEventAfterGenerateIncomingArc<ScType::EdgeAccessConstPosPerm>>
+{
+public:
+  // Here you should specify class of actions which the given agent performs. 
+  // Here `GetActionClass` overrides `GetActionClass` 
+  // in `ScAgent` class. This overriding is required.
+  ScAddr GetActionClass() const override;
+  // Here you should implement program of the given agent. 
+  // This overriding is required.
+  ScResult DoProgram(
+    ScEventAfterGenerateIncomingArc<
+      ScType::EdgeAccessConstPosPerm> const & event, 
+    ScAction & action) override;
```

Comment:
is there somewhere a mention of what is passed as action for different events?

Reply:
Described